File: py/tree.py
class Tree:
    """A basic binary tree"""
    
    def __init__(self):
        self.root = None

    def insert(self, node):
        if (self.root == None):
            self.root = node
        else:
            self.root.insert(node)

    def bulk_insert(self, numbers):
        for i in numbers:
            n = Node(i)
            self.insert(n)

    def report(self):
        if (self.root != None):
            self.root.report()
        else:
            print "No data"

    def contains(self, number):
        if (self.root != None):
            return self.root.contains(number)
        else:
            return false

class Node:
    """A node in the basic binary tree"""

    def __init__ (self, data):
        """Create a node"""
        self.data = data
        self.left = None
        self.right = None

    def insert(self, node):
        if (self.data > node.data):
            if (self.left == None):
                self.left = node
            else:
                self.left.insert(node)
        else:
            if (self.right == None):
                self.right = node
            else:
                self.right.insert(node)

    def report(self):
        if (self.left != None):
            self.left.report()
        print str(self.data)
        if (self.right != None):
            self.right.report()

    def contains(self, number):
        if (self.data == number):
            return True
        else:
            if (self.data > number):
                if (self.left != None):
                    return self.left.contains(number)
                else:
                    return False
            else:
                if (self.right != None):
                    return self.right.contains(number)
                else:
                    return False

if __name__ == '__main__':
    tree = Tree()
    tree.bulk_insert([4, 3, 6, 9, 13, 2, 3, 8, 3, 8, 4])
    tree.report()
    for i in range(1, 15):
        print str(i) + ": " + str(tree.contains(i))
    

File: cpp/tree.cpp

#include "tree.h"
#include <iostream>

using namespace std;

Tree::Tree() {
    root = NULL;
}

void Tree::insert(int num) {
    if (root == NULL) {
        root = new Node(num);
    } else {
        root->insert(num);
    }
}

bool Tree::contains(int num) {
    bool ret = false;
    if (root != NULL) {
        Node* node = root->find(num);
        if (node != NULL) {
            ret = true;
        }
    }
    return ret;
}

int Tree::size() {
  int ret = 0;
  if (root != NULL) {
    ret = root->size();
  }
  return ret;
}

Node* Node::find(int num) {
    Node* ret;
    if (data == num) {
        ret = this;
    } else if (data > num) {
        if (left == NULL) {
            left = new Node(num);
        } else {
            left->insert(num);
        }
        ret = left;
    } else {
        if (right == NULL) {
            right = new Node(num);
        } else {
            right->insert(num);
        }
        ret = right;
    }
    return ret;
}

Node::Node(int num) {
  data = num;
  left = NULL;
  right = NULL;
}

void Node::insert(int num) {
    if (data > num) {
        if (left == NULL) {
            left = new Node(num);
        } else {
            left->insert(num);
        }
    }
    else {
        if (right == NULL) {
            right = new Node(num);
        } else {
            right->insert(num);
        }
    }
}

int Node::size() {
  int ret = 1; // for myself
  if (this->left != NULL) {
    ret = ret + left->size();
  }
  if (right != NULL) {
    ret = ret + right->size();
  }
  return ret;
}
